# 🎧 Projeto: Spotify Hit Analysis

Objetivo do projeto:

Identificar padrões entre os maiores sucessos musicais globais para ajudar uma gravadora a tomar decisões estratégicas sobre produção musical, marketing e investimento em artistas.

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path

In [ ]:
# Leitura do arquivo
path = Path('../data/raw//Spotify Most Streamed Songs.csv')
df = pd.read_csv(path, sep=',')

In [12]:
# Configuração de visualização de Linhas e Colunas

# Mostrar todas as colunas
pd.set_option('display.max_columns', None)

# Mostrar até 200 linhas
pd.set_option('display.max_rows', 200)

## Informações básicas sobre a pista:

### track_name: Nome da música.
- nome_do(s)_artista(s): Nome do(s) artista(s) que interpreta(m) a música.
- artist_count: Número de artistas que contribuíram para a música.
- released_year, released_month, released_day: Detalhes da data de lançamento.

### Métricas de streaming:
- in_spotify_playlists: Número de playlists do Spotify em que a música está presente.
- in_spotify_charts: Posição da música nas paradas do Spotify.
- streams: Número total de reproduções no Spotify.
- in_apple_playlists, in_apple_charts: Presença em playlists e paradas musicais do Apple Music.
- in_deezer_playlists, in_deezer_charts: Presença em playlists e paradas musicais do Deezer.
- in_shazam_charts: Classificação nas paradas do Shazam.

### Atributos musicais:
- bpm: Batidas por minuto, que representam o andamento da música.
- Key: Tom da música.
- Mode: Indica se a música está em modo maior ou menor.
- danceability_%: Adequação da música para dançar.
- valência_%: Positividade do conteúdo musical da canção.
- energy_%: Nível de energia percebido da música.
- acousticness_%: Presença do som acústico na música.
- Instrumentalidade_%: Proporção de conteúdo instrumental na faixa.
- liveness_%: Presença de elementos de performance ao vivo.
- speechiness_%: Quantidade de palavras faladas na música.

# 1. Entendimento dos Dados

In [ ]:
# Dimensão do Dataset
df.shape

(953, 25)

In [17]:
df.columns

Index(['track_name', 'artist(s)_name', 'artist_count', 'released_year',
       'released_month', 'released_day', 'in_spotify_playlists',
       'in_spotify_charts', 'streams', 'in_apple_playlists', 'in_apple_charts',
       'in_deezer_playlists', 'in_deezer_charts', 'in_shazam_charts', 'bpm',
       'key', 'mode', 'danceability_%', 'valence_%', 'energy_%',
       'acousticness_%', 'instrumentalness_%', 'liveness_%', 'speechiness_%',
       'cover_url'],
      dtype='str')

In [18]:
df.head()

,track_name,artist(s)_name,artist_count,released_year,released_month,released_day,in_spotify_playlists,in_spotify_charts,streams,in_apple_playlists,in_apple_charts,in_deezer_playlists,in_deezer_charts,in_shazam_charts,bpm,key,mode,danceability_%,valence_%,energy_%,acousticness_%,instrumentalness_%,liveness_%,speechiness_%,cover_url
0,Seven (feat. Latto) (Explicit Ver.),"Latto, Jung Kook",2,2023,7,14,553,147,141381703,43,263,45,10,826,125,B,Major,80,89,83,31,0,8,4,Not Found
1,LALA,Myke Towers,1,2023,3,23,1474,48,133716286,48,126,58,14,382,92,C#,Major,71,61,74,7,0,10,4,https://i.scdn.co/image/ab67616d0000b2730656d5...
2,vampire,Olivia Rodrigo,1,2023,6,30,1397,113,140003974,94,207,91,14,949,138,F,Major,51,32,53,17,0,31,6,https://i.scdn.co/image/ab67616d0000b273e85259...
3,Cruel Summer,Taylor Swift,1,2019,8,23,7858,100,800840817,116,207,125,12,548,170,A,Major,55,58,72,11,0,11,15,https://i.scdn.co/image/ab67616d0000b273e787cf...
4,WHERE SHE GOES,Bad Bunny,1,2023,5,18,3133,50,303236322,84,133,87,15,425,144,A,Minor,65,23,80,14,63,11,6,https://i.scdn.co/image/ab67616d0000b273ab5c9c...


In [19]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 953 entries, 0 to 952
Data columns (total 25 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   track_name            953 non-null    str  
 1   artist(s)_name        953 non-null    str  
 2   artist_count          953 non-null    int64
 3   released_year         953 non-null    int64
 4   released_month        953 non-null    int64
 5   released_day          953 non-null    int64
 6   in_spotify_playlists  953 non-null    int64
 7   in_spotify_charts     953 non-null    int64
 8   streams               953 non-null    str  
 9   in_apple_playlists    953 non-null    int64
 10  in_apple_charts       953 non-null    int64
 11  in_deezer_playlists   953 non-null    str  
 12  in_deezer_charts      953 non-null    int64
 13  in_shazam_charts      903 non-null    str  
 14  bpm                   953 non-null    int64
 15  key                   858 non-null    str  
 16  mode               

In [ ]:
# Dados Estatísticos do DataFrame
df.describe()

,artist_count,released_year,released_month,released_day,in_spotify_playlists,in_spotify_charts,in_apple_playlists,in_apple_charts,in_deezer_charts,bpm,danceability_%,valence_%,energy_%,acousticness_%,instrumentalness_%,liveness_%,speechiness_%
count,953.000000,953.000000,953.000000,953.000000,953.000000,953.000000,953.000000,953.000000,953.000000,953.000000,953.00000,953.000000,953.000000,953.000000,953.000000,953.000000,953.000000
mean,1.556139,2018.238195,6.033578,13.930745,5200.124869,12.009444,67.812172,51.908709,2.666317,122.540399,66.96957,51.431270,64.279119,27.057712,1.581322,18.213012,10.131165
std,0.893044,11.116218,3.566435,9.201949,7897.608990,19.575992,86.441493,50.630241,6.035599,28.057802,14.63061,23.480632,16.550526,25.996077,8.409800,13.711223,9.912888
min,1.000000,1930.000000,1.000000,1.000000,31.000000,0.000000,0.000000,0.000000,0.000000,65.000000,23.00000,4.000000,9.000000,0.000000,0.000000,3.000000,2.000000
25%,1.000000,2020.000000,3.000000,6.000000,875.000000,0.000000,13.000000,7.000000,0.000000,100.000000,57.00000,32.000000,53.000000,6.000000,0.000000,10.000000,4.000000
50%,1.000000,2022.000000,6.000000,13.000000,2224.000000,3.000000,34.000000,38.000000,0.000000,121.000000,69.00000,51.000000,66.000000,18.000000,0.000000,12.000000,6.000000
75%,2.000000,2022.000000,9.000000,22.000000,5542.000000,16.000000,88.000000,87.000000,2.000000,140.000000,78.00000,70.000000,77.000000,43.000000,0.000000,24.000000,11.000000
max,8.000000,2023.000000,12.000000,31.000000,52898.000000,147.000000,672.000000,275.000000,58.000000,206.000000,96.00000,97.000000,97.000000,97.000000,91.000000,97.000000,64.000000


In [ ]:
# Função para gerar metadados da base
def gerar_metadados(df):

    metadados = pd.DataFrame({
        'Tipo_de_Dados':df.dtypes,
        'Qtd_de_Nulos':df.isnull().sum(),
        'Total_Registros':len(df),
        '%_Nulos':(df.isnull().sum() / len(df)) * 100,
        'Cardinalidade':df.nunique()
    })

    metadados = metadados.sort_values(by='%_Nulos', ascending=False)

    return metadados

In [25]:
gerar_metadados(df)

,Tipo de Dados,Qtd de Nulos,Total_Registros,%_Nulos,Cardinalidade
key,str,95,953,9.96852,11
in_shazam_charts,str,50,953,5.24659,198
artist_count,int64,0,953,0.00000,8
artist(s)_name,str,0,953,0.00000,645
released_month,int64,0,953,0.00000,12
released_day,int64,0,953,0.00000,31
in_spotify_playlists,int64,0,953,0.00000,879
released_year,int64,0,953,0.00000,50
track_name,str,0,953,0.00000,943
streams,str,0,953,0.00000,949


## Análise Inicial do Dataset

A base de dados contém **953 músicas** e **25 variáveis**, combinando informações sobre:

- Metadados das músicas
- Métricas de popularidade nas plataformas de streaming
- Atributos musicais relacionados à composição sonora

Cada linha representa uma música popular nas plataformas de streaming.

---

## Estrutura Geral da Base

- Existem **943 nomes de músicas únicos**, indicando poucas duplicatas.
- Há **645 artistas diferentes**, sugerindo diversidade de artistas na base.
- A maioria das músicas possui **1 ou 2 artistas**, refletindo a tendência de colaborações na indústria musical.
- O número de músicas em **playlists do Spotify apresenta alta variação**, com valores que chegam a mais de **50 mil playlists**, indicando grande desigualdade na distribuição de exposição.
- As métricas de playlists e charts permitem avaliar **estratégias de promoção e descoberta musical nas plataformas**.

---

### Colunas com tipo de dado incorreto

Algumas variáveis que representam valores numéricos estão armazenadas como texto (`str`):
:
- `streams`
- `in_deezer_playlists`
- `in_shazam_charts`

Essas colunas precisarão ser convertidas para **formato numérico** antes das análises.

---

### Valores ausentes

Foram identificados valores nulos em duas variáveis:

| Coluna | Quantidade de nulos |
|------|------|
| `key` | 95 |
| `in_shazam_charts` | 50 |

Esses valores precisarão ser tratados na etapa de **Data Preparation**.

---

## Principais Observações da Exploração Inicial

- A base apresenta **grande diversidade de artistas**, mas com concentração de músicas recentes.
- As métricas de playlists indicam forte influência das **curadorias de streaming na exposição das músicas**.
- As variáveis musicais permitem investigar **quais características sonoras estão associadas a músicas populares**.

---

## Próximos Passos

Na próxima etapa do projeto serão realizadas:

- Conversão de tipos de dados
- Tratamento de valores ausentes
- Criação de novas variáveis para análise
- Exploração das relações entre **atributos musicais e popularidade das músicas**

In [33]:
# Verificando os Nulos do atributo 'key'
df[df['key'].isnull()].head(10)

,track_name,artist(s)_name,artist_count,released_year,released_month,released_day,in_spotify_playlists,in_spotify_charts,streams,in_apple_playlists,in_apple_charts,in_deezer_playlists,in_deezer_charts,in_shazam_charts,bpm,key,mode,danceability_%,valence_%,energy_%,acousticness_%,instrumentalness_%,liveness_%,speechiness_%,cover_url
12,Flowers,Miley Cyrus,1,2023,1,12,12211,115,1316855716,300,215,745,58,"1,021",118,NaN,Major,71,65,68,6,0,3,7,https://i.scdn.co/image/ab67616d0000b27358039b...
17,What Was I Made For? [From The Motion Picture ...,Billie Eilish,1,2023,7,13,873,104,30546883,80,227,95,24,"1,173",78,NaN,Major,44,14,9,96,0,10,3,Not Found
22,I Wanna Be Yours,Arctic Monkeys,1,2013,1,1,12859,110,1297026226,24,98,582,2,73,135,NaN,Minor,48,44,42,12,2,11,3,https://i.scdn.co/image/ab67616d0000b2734ae1c4...
35,Los del Espacio,"Big One, Duki, Lit Killah, Maria Becerra, FMK,...",8,2023,6,1,1150,31,123122413,22,33,34,7,184,120,NaN,Major,81,63,68,11,0,11,4,Not Found
44,Barbie World (with Aqua) [From Barbie The Album],"Nicki Minaj, Aqua, Ice Spice",3,2023,6,23,1117,80,65156199,82,145,65,16,"1,133",144,NaN,Major,77,75,58,52,0,23,25,https://i.scdn.co/image/ab67616d0000b2737e8f93...
46,I Ain't Worried,OneRepublic,1,2022,5,13,8431,76,1085685420,241,127,458,37,332,140,NaN,Major,71,82,81,11,0,6,5,https://i.scdn.co/image/ab67616d0000b273ec96e0...
58,S91,Karol G,1,2023,7,14,525,41,16011326,34,115,39,6,216,128,NaN,Minor,86,42,72,59,0,9,19,https://i.scdn.co/image/ab67616d0000b273890cfb...
59,cardigan,Taylor Swift,1,2020,7,24,7923,29,812019557,106,112,142,4,215,130,NaN,Minor,61,53,58,55,0,27,4,https://i.scdn.co/image/ab67616d0000b27395f754...
124,Por las Noches,Peso Pluma,1,2021,6,11,457,24,330346424,8,116,4,3,2,92,NaN,Major,81,39,60,31,0,7,3,https://i.scdn.co/image/ab67616d0000b273427d9a...
127,Watermelon Sugar,Harry Styles,1,2019,11,17,21915,34,2322580122,437,115,"1,212",12,NaN,95,NaN,Major,55,56,82,12,0,34,5,https://i.scdn.co/image/ab67616d0000b27377fdcf...


In [34]:
# Verificando os nulos do atributo 'in_shazam_charts'
df[df['in_shazam_charts'].isnull()].head(10)

,track_name,artist(s)_name,artist_count,released_year,released_month,released_day,in_spotify_playlists,in_spotify_charts,streams,in_apple_playlists,in_apple_charts,in_deezer_playlists,in_deezer_charts,in_shazam_charts,bpm,key,mode,danceability_%,valence_%,energy_%,acousticness_%,instrumentalness_%,liveness_%,speechiness_%,cover_url
14,As It Was,Harry Styles,1,2022,3,31,23575,130,2513188493,403,198,863,46,NaN,174,F#,Minor,52,66,73,34,0,31,6,https://i.scdn.co/image/ab67616d0000b2732e8ed7...
54,Another Love,Tom Odell,1,2012,10,15,18371,83,1813673666,250,122,"3,394",19,NaN,123,E,Minor,45,13,54,70,0,9,4,https://i.scdn.co/image/ab67616d0000b2731917a0...
55,Blinding Lights,The Weeknd,1,2019,11,29,43899,69,3703895074,672,199,"3,421",20,NaN,171,C#,Major,50,38,80,0,0,9,7,https://i.scdn.co/image/ab67616d0000b2738863bc...
71,Heat Waves,Glass Animals,1,2020,6,28,22543,63,2557975762,386,144,707,28,NaN,81,B,Major,76,53,53,44,0,9,9,https://i.scdn.co/image/ab67616d0000b273712701...
73,Sweater Weather,The Neighbourhood,1,2012,5,14,16413,61,2282771485,166,87,"1,056",1,NaN,124,A#,Major,61,41,81,5,2,10,3,https://i.scdn.co/image/ab67616d0000b2738265a7...
86,Someone You Loved,Lewis Capaldi,1,2018,11,8,17836,53,2887241814,440,125,"1,800",0,NaN,110,C#,Major,50,45,41,75,0,11,3,https://i.scdn.co/image/ab67616d0000b273fc2101...
127,Watermelon Sugar,Harry Styles,1,2019,11,17,21915,34,2322580122,437,115,"1,212",12,NaN,95,NaN,Major,55,56,82,12,0,34,5,https://i.scdn.co/image/ab67616d0000b27377fdcf...
158,Ghost,Justin Bieber,1,2021,3,19,5866,24,1167330737,107,38,95,0,NaN,154,D,Major,61,41,74,21,0,40,6,https://i.scdn.co/image/ab67616d0000b273e6f407...
159,Under The Influence,Chris Brown,1,2019,10,4,3859,26,929964809,133,181,3,0,NaN,117,A,Minor,73,31,69,6,0,11,4,https://i.scdn.co/image/ab67616d0000b2739a494f...
180,Night Changes,One Direction,1,2014,11,17,7124,18,1131090940,60,20,2,0,NaN,120,G#,Major,67,40,52,86,0,12,4,https://i.scdn.co/image/ab67616d0000b273d304ba...


In [30]:
# Verificar se tem registros com os dois atributos nulos
df[df['key'].isnull() & df['in_shazam_charts'].isnull()]

,track_name,artist(s)_name,artist_count,released_year,released_month,released_day,in_spotify_playlists,in_spotify_charts,streams,in_apple_playlists,in_apple_charts,in_deezer_playlists,in_deezer_charts,in_shazam_charts,bpm,key,mode,danceability_%,valence_%,energy_%,acousticness_%,instrumentalness_%,liveness_%,speechiness_%,cover_url
127,Watermelon Sugar,Harry Styles,1,2019,11,17,21915,34,2322580122,437,115,"1,212",12,NaN,95,NaN,Major,55,56,82,12,0,34,5,https://i.scdn.co/image/ab67616d0000b27377fdcf...
395,Space Song,Beach House,1,2015,1,1,17852,4,789753877,69,76,335,0,NaN,147,NaN,Minor,51,62,79,22,13,14,3,https://i.scdn.co/image/ab67616d0000b2739b7190...
504,Out of Time,The Weeknd,1,2022,1,7,3711,0,339659802,49,88,62,0,NaN,93,NaN,Minor,65,82,74,27,0,32,5,https://i.scdn.co/image/ab67616d0000b2734ab252...
506,We Don't Talk About Bruno,"Adassa, Mauro Castillo, Stephanie Beatriz, Enc...",7,2021,11,19,2785,0,432719968,95,89,44,0,NaN,206,NaN,Minor,58,83,45,36,0,11,8,Not Found
531,Happier Than Ever,Billie Eilish,1,2021,7,30,8476,0,1056760045,138,133,283,0,NaN,65,NaN,Major,31,31,24,76,0,14,4,https://i.scdn.co/image/ab67616d0000b2732a038d...
532,Moth To A Flame (with The Weeknd),"The Weeknd, Swedish House Mafia",2,2021,10,22,7495,17,611994237,114,172,284,2,NaN,120,NaN,Minor,56,16,67,0,0,11,4,Not Found
554,Peaches (feat. Daniel Caesar & Giveon),"Justin Bieber, Daniel Caesar, Giveon",3,2021,3,19,14140,0,1445941661,231,52,612,6,NaN,90,NaN,Major,63,49,68,38,0,42,18,https://i.scdn.co/image/ab67616d0000b273e6f407...
560,Life Goes On,Oliver Tree,1,2020,7,17,2868,0,501541661,43,15,116,0,NaN,80,NaN,Major,70,57,49,19,0,12,8,https://i.scdn.co/image/ab67616d0000b2731c5a48...
727,Somebody That I Used To Know,"Gotye, Kimbra",2,2011,1,1,42798,0,1457139296,217,136,"6,508",1,NaN,129,NaN,Major,86,75,52,54,0,10,4,https://i.scdn.co/image/ab67616d0000b273e1d47c...


In [43]:
# Verificar distribuição dos nulos por ano
df[df['key'].isnull() | df['in_shazam_charts'].isnull()]['released_year'].value_counts().sort_index(ascending=False)

released_year
2023    16
2022    43
2021    24
2020    14
2019     9
2018     1
2017     4
2016     1
2015     2
2014     1
2013     3
2012     4
2011     1
2010     1
2007     1
2004     1
2000     1
1995     1
1994     1
1984     2
1975     1
1968     1
1958     1
1957     1
1952     1
Name: count, dtype: int64

## Análise de Padrão de Valores Nulos

Após identificar valores ausentes nas variáveis `key` e `in_shazam_charts`, foi realizada uma análise exploratória para verificar possíveis padrões nesses registros.

Variável `key`

- A variável `key` representa o **tom musical da faixa**.
- A análise das amostras mostrou que os valores ausentes ocorrem em músicas de diferentes artistas, anos e níveis de popularidade. Inclusive, músicas extremamente populares como *Flowers* (Miley Cyrus) e *I Wanna Be Yours* (Arctic Monkeys) apresentam valores nulos nessa coluna.
- Esse comportamento sugere que os valores ausentes estão associados a **limitações na extração automática da tonalidade musical**, e não a algum padrão específico de negócio.

---

Variável `in_shazam_charts`

- A variável `in_shazam_charts` indica a presença da música nas paradas do Shazam.
- Entretanto, foram observados valores nulos em músicas altamente populares, o que indica que os valores ausentes provavelmente representam **ausência de registro da posição no dataset**, e não necessariamente ausência de popularidade.

---

### Distribuição Temporal dos Nulos

A análise da distribuição por ano revelou maior ocorrência de valores ausentes em músicas mais recentes.

Possíveis explicações incluem:

- dados ainda em atualização no momento da coleta
- metadados musicais ainda não completamente catalogados
- atraso na captura de métricas pelas plataformas

---

### Conclusão

Não foi identificado um padrão claro de negócio associado aos valores nulos. Os registros ausentes parecem ser resultado de limitações na coleta ou processamento dos dados, e não indicam inconsistências críticas no dataset.

# 2. Preparação dos Dados